tmk_core/common/chibios/eeprom_stm32f4.c

```diff
@@ -0,0 +1,342 @@
+/**
+ * @file eeprom_stm32f4.c
+ * @author astro
+ */
+
+#include <string.h>
+#include "eeprom_stm32.h"
+#include "hal.h"
+
+#define EE_SECTOR_ID 3
+#define EE_SECTOR_SIZE 0x4000
```

Comment:
some pointers to documentation where you got these magic values from would be great too. thanks.

Reply:
description for this was added

Reply:
Doesn't this limit the executable flash space (starting at `0x08000000`) to 48kB unless each stm32f4 project's ldscripts get modified?
If they're not changed, then this eeprom implementation will potentially overwrite program code, will it not?

Reply:
Yes, this limits the flash space, currently the qmk need about 30K flash(with all rgb lights enabled). The flash sector was not uniform distributed as the F1 or F3. the sector 0-3 are 16KB each, sector 4 has 64KB, sector 5 to last were 128KB. maybe we can use the last sector and use the DataBuf for caches.

Reply:
Unsure what you're including in your "normal" builds, but I have an ARM board that uses RGB, OLED, external EEPROM and custom matrix -- it's 52kB of flash used. I'm currently working on one that includes colour images for an LCD -- 30kB would get used very quickly.

Now, I'm not targeting F4 for this build, but I don't feel like we should be artifically placing a restriction on the amount of flash available in general. The assertion that around 30kB is "normal" doesn't really hold true now, and certainly won't do so in the future.

Reply:
Normally I'll flash an small bootloader for IAP at the start of flash space. And the qmk firmware will sit behind the bootloader and the emulated eeprom area. This can remove the hard limit of flash space for application. Anyway the latest codes have used the last flash sector for eeprom emulation, so the qmk can programmed at start of the flash for common usage.